File: Java Solutions/SlidingWindow/src/main/java/org/example/LongestSubstringWithoutRepeating.java
package org.example;

import java.util.HashMap;

public class LongestSubstringWithoutRepeating {
    public int lengthOfLongestSubstring(String s) {
        if (s.isEmpty()){
            return 0;
        }
        int max_length = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        for (int left = 0, right = 0; right < s.length(); right++) {
            char letter = s.charAt(right);
            if (map.containsKey(letter) && map.get(letter) >= left) {
                left = map.get(letter) + 1;
                map.put(letter, right);
            }
            map.put(letter, right);
            max_length = Integer.max(max_length, right - left);
        }
        return max_length + 1;
    }
}

File: Python Solutions/sliding_window/longest_substring_without_repeating.py
def length_of_longest_substring(s: str) -> int:
    if not s:
        return 0
    used_letters = {}
    left = 0
    length = 0
    
    for right, letter in enumerate(s):
        if letter in used_letters and used_letters[letter] >= left:
            left = used_letters[letter] + 1
            used_letters[letter] = right
        used_letters[letter] = right
        length = max(length, right - left)

    return length + 1
